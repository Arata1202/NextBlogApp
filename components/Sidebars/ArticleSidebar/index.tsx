'use client';

import React from 'react';
import Search from '../Elements/Search';
import Profile from '../Elements/Profile';
import Category from '../Elements/Category';
import Tag from '../Elements/Tag';
import Archive from '../Elements/Archive';
import Popular from '../Elements/Popular';
import Recent from '../Elements/Recent';
import Display from '../../Adsense/Display';
import { Article } from '@/libs/microcms';
import TableOfContents from '../../Articles/Elements/TableOfContent';
import { useEffect, useState } from 'react';
import { useGuardObserver } from '@/hooks/MutationObserver';

type Props = {
  articles?: Article[];
  contentBlocks?: { rich_text2?: string }[];
};

interface ContentBlock {
  rich_text2?: string;
}

interface Heading {
  id: string;
  title: string;
  level: number;
}

function useExtractHeadings(contentBlocks: ContentBlock[]): Heading[] {
  const [headings, setHeadings] = useState<Heading[]>([]);

  useEffect(() => {
    const extractedHeadings: Heading[] = [];

    contentBlocks.forEach((block) => {
      if (block.rich_text2) {
        const tempDiv = document.createElement('div');
        tempDiv.innerHTML = block.rich_text2;
        const blockHeadings = Array.from(tempDiv.querySelectorAll('h1, h2, h3, h4, h5, h6')).map(
          (el) => ({
            id: el.id,
            title: el.textContent || '',
            level: parseInt(el.tagName[1], 10),
          }),
        );
        extractedHeadings.push(...blockHeadings);
      }
    });

    setHeadings(extractedHeadings);
  }, [contentBlocks]);

  return headings;
}

export default function ArticleSidebar({ articles, contentBlocks = [] }: Props) {
  const headings = useExtractHeadings(contentBlocks);
  useGuardObserver();

  return (
    <div className="lg:col-span-1 lg:w-full lg:h-full mut-guard">
      <div className="sidebar">
        <Search />
        <Profile />
        <div className="FirstAd mt-5 mut-guard">
          <Display slot="8452341403" />
        </div>
        <Category />
        <Tag />
        <div className="FirstAd mt-5 mut-guard">
          <Display slot="9574685533" />
        </div>
        <Archive />
        <Popular />
        <Recent articles={articles} />
      </div>
      <div className="SidebarTableOfContens mobile">
        {headings.length > 0 && <TableOfContents headings={headings} />}
        <a href="https://www.buymeacoffee.com/realunivlog" target="_blank">
          <img
            className="mt-5 m-auto hover:opacity-60"
            loading="lazy"
            src="https://cdn.buymeacoffee.com/buttons/v2/default-yellow.png"
            width="160"
          />
        </a>
        <div className="text-center mt-4" style={{ fontSize: '12px' }}>
          もしこの記事が役に立ったなら、
          <br />
          こちらから ☕ を一杯支援いただけると喜びます
        </div>
      </div>
    </div>
  );
}
