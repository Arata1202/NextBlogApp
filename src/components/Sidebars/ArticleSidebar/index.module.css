.BuyMeaCoffee {
  font-size: 12px;
}
