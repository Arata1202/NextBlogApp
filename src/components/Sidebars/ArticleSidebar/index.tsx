'use client';

import { useEffect, useState } from 'react';
import { Article } from '@/libs/microcms';
import styles from './index.module.css';
import Display from '../../Common/Adsense/Display';
import Search from '../Elements/Search';
import Profile from '../Elements/Profile';
import Category from '../Elements/Category';
import Tag from '../Elements/Tag';
import Archive from '../Elements/Archive';
import Popular from '../Elements/Popular';
import Recent from '../Elements/Recent';
import TableOfContents from '../../Common/TableOfContent';
import { useGuardObserver } from '@/hooks/MutationObserver';

type Props = {
  articles: Article[];
  contentBlocks: { rich_text?: string }[];
};

interface ContentBlock {
  rich_text?: string;
}

interface Heading {
  id: string;
  title: string;
  level: number;
}

function useExtractHeadings(contentBlocks: ContentBlock[]): Heading[] {
  const [headings, setHeadings] = useState<Heading[]>([]);

  useEffect(() => {
    const extractedHeadings: Heading[] = [];

    contentBlocks.forEach((block) => {
      if (block.rich_text) {
        const tempDiv = document.createElement('div');
        tempDiv.innerHTML = block.rich_text;
        const blockHeadings = Array.from(tempDiv.querySelectorAll('h1, h2, h3, h4, h5, h6')).map(
          (el) => ({
            id: el.id,
            title: el.textContent || '',
            level: parseInt(el.tagName[1], 10),
          }),
        );
        extractedHeadings.push(...blockHeadings);
      }
    });

    setHeadings(extractedHeadings);
  }, [contentBlocks]);

  return headings;
}

export default function ArticleSidebar({ articles, contentBlocks }: Props) {
  const headings = useExtractHeadings(contentBlocks);
  useGuardObserver();

  return (
    <>
      <div className="lg:col-span-1 lg:w-full lg:h-full mut-guard">
        <Search />
        <Profile />
        <div className="FirstAd mt-5 mut-guard">
          <Display slot="8452341403" />
        </div>
        <Category />
        <Tag />
        <div className="FirstAd mt-5 mut-guard">
          <Display slot="9574685533" />
        </div>
        <Archive />
        <Popular />
        <Recent articles={articles} />
        <div className="SidebarTableOfContens mobile">
          {headings.length > 0 && <TableOfContents headings={headings} />}
          <a href="https://www.buymeacoffee.com/realunivlog" target="_blank">
            <img
              className="mt-5 m-auto hover:opacity-60"
              loading="lazy"
              src="https://cdn.buymeacoffee.com/buttons/v2/default-yellow.png"
              width="160"
            />
          </a>
          <div className={`${styles.BuyMeaCoffee} text-center mt-4`}>
            もしこの記事が役に立ったなら、
            <br />
            こちらから ☕ を一杯支援いただけると喜びます
          </div>
        </div>
      </div>
    </>
  );
}
