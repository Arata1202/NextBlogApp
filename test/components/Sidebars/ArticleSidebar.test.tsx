import { render } from '@testing-library/react';
import ArticleSidebar from '@/components/Sidebars/ArticleSidebar';
import type { Article } from '@/libs/microcms';

jest.mock('@/components/Sidebars/Elements/Search');
jest.mock('@/components/Sidebars/Elements/Profile');
jest.mock('@/components/Sidebars/Elements/Category');
jest.mock('@/components/Sidebars/Elements/News');
jest.mock('@/components/Sidebars/Elements/Tag');
jest.mock('@/components/Sidebars/Elements/Archive');
jest.mock('@/components/Sidebars/Elements/Popular');
jest.mock('@/components/Sidebars/Elements/Recent');
jest.mock('@/components/Adsense/Display');
jest.mock('@/components/Articles/Elements/TableOfContent');

const mockArticles: Article[] = [
  {
    id: '1',
    title: 'サンプル記事1',
    description: 'これはサンプル記事1の説明です',
    thumbnail: { url: 'https://example.com/thumbnail1.jpg' },
    publishedAt: '2023-10-30T12:00:00Z',
    content_blocks: [],
    introduction_blocks: [],
    createdAt: '2023-10-30T12:00:00Z',
    updatedAt: '2023-10-30T12:00:00Z',
  },
  {
    id: '2',
    title: 'サンプル記事2',
    description: 'これはサンプル記事2の説明です',
    thumbnail: { url: 'https://example.com/thumbnail2.jpg' },
    publishedAt: '2023-10-30T12:00:00Z',
    content_blocks: [],
    introduction_blocks: [],
    createdAt: '2023-10-30T12:00:00Z',
    updatedAt: '2023-10-30T12:00:00Z',
  },
];

const mockContentBlocks = [
  {
    rich_text2: 'サンプルヘッディング',
    adsense: '1234567890',
    bubble_name: '名前',
    bubble_text: 'テキスト',
    bubble_image: { url: 'https://example.com/image.jpg' },
  },
];

describe('ArticleSidebar', () => {
  test('スナップショット（ArticleSidebar）', () => {
    const { asFragment } = render(
      <ArticleSidebar articles={mockArticles} contentBlocks={mockContentBlocks} />,
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
