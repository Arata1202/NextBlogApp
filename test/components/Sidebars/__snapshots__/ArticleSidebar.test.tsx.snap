// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArticleSidebar スナップショット（ArticleSidebar） 1`] = `
<DocumentFragment>
  <div
    class="lg:col-span-1 lg:w-full lg:h-full"
  >
    <div
      class="sidebar"
    >
      <div
        class="FirstAd pc mt-5"
      />
    </div>
    <div
      class="SidebarTableOfContens mobile"
    />
  </div>
</DocumentFragment>
`;
